test/hotspot/jtreg/serviceability/jvmti/vthread/HeapDump/VThreadInHeapDump.java

```diff
@@ -0,0 +1,285 @@
+            // If something goes wrong with heap dumping most likely we'll get crash of the target VM.
+            while (!p.waitFor(5, TimeUnit.SECONDS)) {
+                if (!theApp.getProcess().isAlive()) {
+                    log("ERROR: target VM died, killing jcmd...");
+                    p.destroyForcibly();
+                    throw new Exception("Target VM died");
+                }
+            }
+
+            if (p.exitValue() != 0) {
+                throw new Exception("Jcmd exited with code " + p.exitValue());
+            }
+        } finally {
+            LingeredApp.stopApp(theApp);
+        }
+    }
+
+    private static void verifyDump(File dumpFile) throws Exception {
+        Asserts.assertTrue(dumpFile.exists(), "Heap dump file not found.");
+
+        log("Reading " + dumpFile + "...");
+        try (Snapshot snapshot = Reader.readFile(dumpFile.getPath(), true, 0)) {
+            log("Resolving snapshot...");
+            snapshot.resolve(true);
+            log("Snapshot resolved.");
+
+            // Log all threads with stack traces and stack references.
+            List<ThreadObject> threads = snapshot.getThreads();
+            List<Root> roots = Collections.list(snapshot.getRoots());
+            log("Threads:");
+            for (ThreadObject thread: threads) {
+                StackTrace st = thread.getStackTrace();
+                StackFrame[] frames = st.getFrames();
+                log("thread " + thread.getIdString() + ", " + frames.length + " frames");
+
+                List<Root> stackRoots = findStackRoot(roots, thread);
+                for (int i = 0; i < frames.length; i++) {
+                    log("  - [" + i + "] "
+                        + frames[i].getClassName() + "." + frames[i].getMethodName()
+                        + frames[i].getMethodSignature()
+                        + " (" + frames[i].getSourceFileName()
+                        + ":" + frames[i].getLineNumber() + ")");
+
+                    for (Root r: stackRoots) {
+                        StackFrame[] rootFrames = r.getStackTrace().getFrames();
+                        // the frame this local belongs to
+                        StackFrame frame = rootFrames[rootFrames.length - 1];
+                        if (frame == frames[i]) {
+                            JavaHeapObject obj = snapshot.findThing(r.getId());
+                            JavaClass objClass = obj.getClazz();
+                            log("      " + r.getDescription() + ": " + objClass.getName());
+                        }
+                    }
+                }
+            }
+
+            // Verify objects from thread stacks are dumped.
+            test(snapshot, VThreadInHeapDumpTarg.VThreadMountedReferenced.class);
+            test(snapshot, VThreadInHeapDumpTarg.PThreadReferenced.class);
+            //test(snapshot, VThreadInHeapDumpTarg.VThreadUnmountedReferenced.class);
```

Comment:
It'd be nice to add a comment explaining why this line was commented out.

Reply:
added